Ion-java-cli-process

cheqianh opened the pull request:
### **Description**
Ion-java-cli has 2 core parts for version one: process and compare.
This commit is for process which supports: 
- --output
- --output-format
- --error-report

note: I just noticed some check style errors such as I forget space after if-else statement, I'll fix them in the next commit.

### **What I finished**
1. Support binary/text/pretty format
test: some example files which include string value, integer, list, structure and local symbol tables.
2. Support events format
test: some example files which include string value, integer, list, structure and local symbol tables.
3. Support -o options which outputs a file to ion-java/.
test: some example files which include string value, integer, list, structure and local symbol tables.
4. Support -e options which outputs error report to ion-java/. 
test: some example files which include string value, integer, list, structure and local symbol tables.
5. Implement Event structure
6. Able to Write a Event structure to outputStream in Events type
test: converted a ionReader value to a Scalar event
7. Convert ionStream to Event structure
test: converted five types of events to outputStream(container_start, container_end, symbol_table, scalar and stream_end)
8. Check if Symbol table changes
test: It works for two local symbol tables with **the same imports** or **the second local symbol table imports the previous local symbol table's name (which means the imports field of the current symbol table is previous symbol table's name)**

### **What I'm going to do next/not implement yet** 
1. Error report structure
2. Embedded stream
3. --output-format none

### **Question/Specific areas what to check:**
1. If Annotation is $0, both text and import location show 'null.string' while ion-c just simply shows 'null'. Not sure if it's because ion-java and ion-c performance different. 
Reproduce/code: Code is in Event.java line 55-59

2. The output file is always start with a blank line. 
Reproduce/code: To reproduce this, use -o <fileName> and -f events options and generate a new output file which its first line is blank.

3. When checking if 2 symbol tables are same, my logic checks the cases of systemTables, shareTables and substitue too which may be redundant. 
reproduce/code: Code is in IonProcess.java: line 236

4. Local symbol table has 2 fields, imports and symbols. I only checked imports when checking if 2 symbol tables are same, Do we need to check symbols as well?)
reproduce/code: Code is in IonProcess.java: line 236

cheqianh committed:
working on process (events type)

almost finish event type

ready for first pr

tgregg commented on pom.xml:
```diff
@@ -82,6 +82,17 @@
+    </dependency>
+    <dependency>
+      <groupId>com.github.luben</groupId>
+      <artifactId>zstd-jni</artifactId>
```
It looks like this dependency isn't used. Is that correct?

cheqianh replied:
removed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+                File myFile = new File(outputFileName);
+                FileOutputStream out = new FileOutputStream(myFile);
+                outputStream = new BufferedOutputStream(out, BUFFER_SIZE);
+            }
```
Let's structure this section so that you only assign to `outputStream` once. That way we make sure that every `OutputStream` can be closed once.

cheqianh replied:
I removed the previous code and write a helper function (ionProcess.java line 170-193)
Now I just use a single line to declare outputStream (IonProcess.java line 60)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+                FileOutputStream out = new FileOutputStream(myFile);
+                errorReportOutputStream = new BufferedOutputStream(out, BUFFER_SIZE);
+            }
+        }
```
Same comment here about assigning `errorReportOutputStream`.

cheqianh replied:
Same as above, ionProcess.java line 62

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+            try (
+                    InputStream inputStream = new BufferedInputStream(new FileInputStream(path));
+            ) {
+                IonReader ionReader = IonReaderBuilder.standard().build(inputStream);
```
IonReader is `Closeable`, so it should go inside the try-with-resources as well.

cheqianh replied:
moved (IonPorcess line 83)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+        }
+
+        try {
+            ionWriterForOutput.finish();
+            ionWriterForErrorReport.finish();
```
You might as well put these two lines inside the previous try block, after `process`. The behavior of these lines is undefined if the previous block throws an exception anyway.

cheqianh replied:
Moved, (which is currently in IonProcess line 87-88)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+    private static void process(IonWriter ionWriterForOutput,IonWriter ionWriterForErrorReport, IonReader ionReader) throws Exception {
+        while(ionReader.next() != null){
+            ionWriterForOutput.writeValue(ionReader);
+        }
```
You can use `ionWriterForOutput.writeValues(ionReader)` to achieve the same.

Where do you write the error report?

cheqianh replied:
Solved in the third commit (IonProcess.java 98)

By this way, it print all the symbol tables on pretty/text format while ion-c only print the first local symbol table. Is it expected?

tgregg replied:
> By this way, it print all the symbol tables on pretty/text format while ion-c only print the first local symbol table. Is it expected?

Can you give me an example so I can understand what you're referring to?

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+            }
+        } catch (IOException e) {
+            System.err.println("Failed to close OutputStream: " + e.getMessage());
+        }
```
I recommend combining several of these try/catch blocks into one. The slight differences in the messages aren't really gaining you anything, and the exceptions are not recoverable, so you cannot move on to the next block if the previous one throws an exception anyway.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+        }
+    }
+
+    private static void process(IonWriter ionWriterForOutput,IonWriter ionWriterForErrorReport, IonReader ionReader) throws Exception {
```
I think `Exception` is too general here. `IonWriter.writeValue` can throw `IOException`, so let's declare that.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+        }
+    }
+
+    private static void processFilesInEvents(IonWriter ionWriterForOutput,
```
I recommend removing all the catch blocks from this method and just allowing any exceptions to bubble up. They can be handled somewhere up the call stack.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/OutputFormat.java:
```diff
@@ -0,0 +1,76 @@
+
+    public String getName() {
+        return name;
+    }
```
You can delete all of this and just use `OutputFormat.valueOf(name.toUpperCase())`. For example, `OutputFormat binary = OutputFormat.valueOf("binary".toUpperCase());`

cheqianh replied:
two follow-ups after deleting this:

1. Since we simply convert all letters of the string to upper case, "event" performances the same as "EVENTS" or "eVeNtS".

2. ValueOf(name.toUpperCase()) method will throw an IllegalArgumentException if the name exist in enum. 

To solve issue 1, I kept the name field and compare it with the parsed argument to make sure they are matched.(OutputFormat.java 53-59)

There are two ways to solve issue 2, catch IllegalArgumentException or create a help function to check if the name exists in enum. I created a function called nameToOutputFormat (OutputFormat.java 63-70) which checks if the name exist and throw a CmdLineException if the name is not exist in enum.

tgregg replied:
I don't really see issue 1 as a problem. You can accept all capitalizations of the same word the same way. I think having a separate "name" field is redundant because enums already have this.

For issue 2, I think just allowing IllegalArgumentException to bubble up is fine. Your solution to handle the exception is also fine.

cheqianh replied:
Removed name field, and catch IllegalArgumentException in main now. (IonProcess.java 57)

tgregg commented on test/tools/cli/IonWriterCreator.java:
```diff
@@ -0,0 +1,13 @@
+ * given only the OutputStream destination.
+ */
+public interface IonWriterCreator {
+    IonWriter createIonWriter(OutputStream outputStream);
```
If you only use this in OutputFormat, you can just add `createIonWriter` as an abstract method in OutputFormat to achieve the same effect -- no need for a top-level interface.

cheqianh replied:
Deleted and the new function is in (OutputFormat.java 51)

tgregg commented on test/tools/events/ImportDescriptor.java:
```diff
@@ -0,0 +1,45 @@
+public class ImportDescriptor {
+    private String import_name;
+    private int max_id;
+    private int version;
```
To conform to Java style, let's use camel case (importName, maxId, version).

cheqianh replied:
Changed in the third commit.

tgregg commented on test/tools/events/EventType.java:
```diff
@@ -0,0 +1,10 @@
+    SCALAR,
+    SYMBOL_TABLE,
+    STREAM_END,
+    UNKNOWN
```
What is UNKNOWN used for? We should probably remove it.

cheqianh replied:
Deleted

tgregg commented on test/tools/events/Event.java:
```diff
@@ -0,0 +1,95 @@
+    public SymbolToken field_name;
+    public SymbolToken[] annotations;
+    public String value_text;
+    public byte[] value_binary;
```
Let's use camel case to conform to Java style.

Also, it looks like all of these fields are only set in the constructor, so they should be declared `final`.

cheqianh replied:
Changed (Event.java 12-15)

tgregg commented on test/tools/events/ImportDescriptor.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void setVersion(int version) {
+        this.version = version;
+    }
```
Are these setters used? If not, they should be removed and the underlying fields should be declared `final`, as they are only set in the constructor.

cheqianh replied:
Deleted

tgregg commented on test/tools/events/Event.java:
```diff
@@ -0,0 +1,95 @@
+        ionWriterForOutput.stepIn(IonType.STRUCT);
+        if(this.eventType != null){
+            ionWriterForOutput.setFieldName("event_type");
+            ionWriterForOutput.writeString(this.eventType.toString());
```
Use `writeSymbol`.

cheqianh replied:
Changed (Event.java 35)

tgregg commented on test/tools/events/Event.java:
```diff
@@ -0,0 +1,95 @@
+        }
+        if(this.ionType != null){
+            ionWriterForOutput.setFieldName("ion_type");
+            ionWriterForOutput.writeString(this.ionType.toString());
```
Use `writeSymbol`.

cheqianh replied:
Changed (Event.java 39)

tgregg commented on test/tools/events/Event.java:
```diff
@@ -0,0 +1,95 @@
+                ionWriterForOutput.writeString(text);
+                if(text == null) {
+                    ionWriterForOutput.setFieldName("import_location");
+                    ionWriterForOutput.writeString(null);
```
Use `writeNull` instead of `writeString`; if the value were present, it would be a struct, not a string.

cheqianh replied:
changed (Event.java 58)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+            try (
+                    InputStream inputStream = new BufferedInputStream(new FileInputStream(path));
+            ) {
+                IonReader ionReader = IonReaderBuilder.standard().build(inputStream);
```
Move the reader instantiation into the try()

cheqianh replied:
Moved

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+     * this function is only used for CONTAINER_START and SCALAR eventType since other types don't need initial data.
+     */
+    private static Event ionStreamToEvent(EventType eventType, IonReader ionReader) throws Exception {
+        if(ionReader.getType() == null) return null;
```
It seems like this is actually an illegal state. Should you throw `IllegalStateException` in this case, or simply state that callers must provide a reader that is positioned on a value?

cheqianh replied:
I throw an IllegalStateException(IonProcess.java 200)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+                    if(!isSameSymbolTable(x_table[i], y_table[i]))  return false;
+                }
+            }
+            return true;
```
In addition to the imports, you also need to check the local symbols lists.

cheqianh replied:
Checked (IonProcess.java 244-255)

Questions: 
1. I tried some examples on ion-c, it treats local symbol tables with the same import as same no matter if they have different symbols.

2. What is symbol's scope? Are they accumulated every time pass a local symbol table or reset when they encounter a new symbol table?

tgregg replied:
1. Right, you only really need to create a SYMBOL_TABLE event when the imports change. However in ion-c, there's a nice way of determining when this happens by registering a callback function with the reader, whereas in ion-java there is no similar mechanism (short of using a low-level reader and manually handling the symbol tables, which we'd like to avoid). So we're left needing to detect symbol table changes by comparing the symbol tables. I do think you may be able to simplify this by comparing just the imports of the two symbol tables and creating a SYMBOL_TABLE event when they are different.

2. A symbol table is in scope until either a) an Ion version marker is encountered, or b) a new non-append local symbol table is encountered. Note: appended local symbol tables add new symbols without resetting the previously-declared symbols. It is also not possible to append imports, so appended local symbol tables never need a SYMBOL_TABLE event.

cheqianh replied:
Thank you for the answer, I'll check symbols (IonProcess.java 244-255)

tgregg replied:
Sorry, the conclusion I meant to to convey was this:

> I do think you may be able to simplify this by comparing just the imports of the two symbol tables and creating a SYMBOL_TABLE event when they are different.

This would allow you to simplify the current logic by not comparing local symbols. I recommend trying that and comparing it to ion-c's behavior. If they match, let's do it.

cheqianh replied:
Sure, I'll only check imports.

it performances same as ion-c now.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+                }
+            }
+            return true;
+        } else if (x.isSubstitute() && y.isSubstitute()){
+            //TODO
+            return true;
```
I think you can drop this branch. I don't believe it can be hit anyway, as substitute symbol tables are always shared symbol tables.

cheqianh replied:
deleted

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,314 @@
+                        + "events | none). 'events' is only available with the\n"
+                        + "'process' command, and outputs a serialized EventStream\n"
+                        + "representing the input Ion stream(s).")
+        private String outputFormatName = OutputFormat.PRETTY.getName();;
```
Superfluous semicolon at the end.

cheqianh replied:
deleted

tgregg reviewed (commented):
Good start!

I think I've addressed questions 1, 3, and 4 with comments. As for question 2, it doesn't sound like a big deal, but you may be able to debug it to figure out why it happens.

cheqianh committed:
pr1 code review revision 2

cheqianh committed:
revision 2 fix some bugs

cheqianh commented:
Additional question:
1. I checked styles such as extra space, camel format. if I miss anything please comment the line number, I"ll fix them in the next commit :)
2. Not sure if I named functions and variables appropriately. For example, In IonProcess.java 228, I named two symbol tables x and y.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,312 @@
+import tools.events.ImportDescriptor;
+
+
+import java.io.*;
```
Let's list every import individually, not using `*`.

cheqianh replied:
Changed, My IDE did this automatically once I have 5+ imports from the same package.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,312 @@
+
+                ionWriterForOutput.finish();
+                ionWriterForErrorReport.finish();
+            } catch (IOException e) {
```
Since you catch IOException in main(), you no longer need to catch it here. Just declare that this method throws IOException.

cheqianh replied:
I handled most of the Exceptions in main now.

Question:
I moved OutputStream of output and errorReport inside try-with-resource now (IonProcess 62-65)

I noticed that try..catch method doesn't catch exceptions from the try block (IonProcess.java 70-74). For example, if processFilesInEvents method (IonProcess.java 73) throw an IOException, **catch(IOException e) { ... }** part (IonProcess.java line 75) isn't called.

I found that the reason of it is because of declaration of errorReportOutputStream (IonProcess.java line 65), if I moved this line out of try-with-resource, it works normally. 

Do you want me to move it outside of the try with resources or there is a better way to solve this issue?

tgregg replied:
I'm confused. If line 71 or 73 throws an IOException, it should definitely be caught by the catch on line 75. This shouldn't have anything to do with line 65. What am I missing here?

cheqianh replied:
I think I know the reason now.

tgregg replied:
I just ran this snippet and `"Failed to close OutputStream: test"` does indeed get printed. I think there may be something else going on here.

cheqianh replied:
Thanks for testing, I guess I found the reason: can't build an ionWriter with a null outputstream which will throw a NullPointerException. I'll fix it.

tgregg commented on test/tools/events/ImportDescriptor.java:
```diff
@@ -0,0 +1,27 @@
+        return version;
+    }
+
+    final int getMax_id() {
```
The method names should also be camel case (no underscores).

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,312 @@
+                        break;
+                }
+            }
+        } catch (IOException e ) {
```
I recommend removing the try/catch from this method and moving the invocations of this method into the try-with-resources of the caller, which already handles IOException.

cheqianh replied:
moved

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -0,0 +1,312 @@
+            SymbolTable[] yTable = y.getImportedTables();
+            //compare imports
+            if (xTable.length == yTable.length) {
+                for (int i=0; i < xTable.length; i++) {
```
Please always put a space around an `=`, i.e. `i = 0`

cheqianh replied:
Changed, I'll keep these styles in mind.

Do we have any tools or IDE plugin that checks styles in case I miss them?

tgregg replied:
We don't have a standard style configuration for IntelliJ right now, but you can update your settings individually to enforce these things I think.

cheqianh replied:
Got it Thanks!

tgregg reviewed (commented):
Good progress.

> Not sure if I named functions and variables appropriately. For example, In IonProcess.java 228, I named two symbol tables x and y.

I think this is fine in this case.

cheqianh committed:
fix camel style error

fixing style bugs

code review revision 3 fixed some style bug

finished rivision 3

cheqianh commented:
I created another pr2 whose destination is pr1: https://github.com/cheqianh/ion-java/pull/2    

Thanks!

tgregg reviewed (approved):
Approving this one. Any further changes can be made in subsequent CRs. Nice job!
